### PR TITLE
feat: cover registration logic

### PR DESCRIPTION
**Pull request by arein**
# Description

Covers registration logic in functional test.

Towards #13 

## How Has This Been Tested?

`cargo test`

## Due Diligence

* [ ] Breaking change
* [ ] Requires a documentation update
* [ ] Requires a e2e/integration test update

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -169,7 +169,7 @@ pub async fn bootstap(mut shutdown: broadcast::Receiver<()>, config: Config) ->
                 received_notifications: notification_counter,
             },
         )
-    } else {
+    } else if state.config.environment != "test" {
```
Comment (HarryET) [inline src/lib.rs]: not needed please use `cfg!`

File: src/stores/client.rs [hunk 2]
```diff
@@ -60,15 +60,17 @@ impl ClientStore for sqlx::PgPool {
...
             sqlx::QueryBuilder::new("DELETE FROM public.notifications WHERE client_id = ");
         notification_query_builder.push_bind(id);
-        notification_query_builder.push("and tenant_id = ");
+        notification_query_builder.push(" and tenant_id = ");
         notification_query_builder.push_bind(tenant_id);
         let notification_query = notification_query_builder.build();
 
         self.execute(notification_query).await?;
 
+        println!("Deleting client {}", id);
```
Comment (HarryET) [inline src/stores/client.rs]: Please remove debug lines or use `tracing`

File: src/stores/client.rs [hunk 3]
```diff
@@ -60,15 +60,17 @@ impl ClientStore for sqlx::PgPool {
...
+
         let mut query_builder = sqlx::QueryBuilder::new("DELETE FROM public.clients WHERE id = ");
         query_builder.push_bind(id);
-        query_builder.push("and tenant_id = ");
+        query_builder.push(" and tenant_id = ");
```
Comment (HarryET) [inline src/stores/client.rs:74]: Why spaces?

File: src/stores/client.rs [hunk 4]
```diff
@@ -60,15 +60,17 @@ impl ClientStore for sqlx::PgPool {
         let mut notification_query_builder =
             sqlx::QueryBuilder::new("DELETE FROM public.notifications WHERE client_id = ");
         notification_query_builder.push_bind(id);
-        notification_query_builder.push("and tenant_id = ");
+        notification_query_builder.push(" and tenant_id = ");
```
Comment (HarryET) [inline src/stores/client.rs:66]: why spaces?

**Review by HarryET — CHANGES_REQUESTED** [review]

File: src/stores/client.rs (hunk 2 above)
Comment (arein) [reply src/stores/client.rs]: ups. thank you!

File: src/stores/client.rs (hunk 3 above)
Comment (arein) [reply src/stores/client.rs:74]: did this work before? It failed unless there was a space for me.

File: src/stores/client.rs (hunk 4 above)
Comment (arein) [reply src/stores/client.rs:66]: see above

File: src/lib.rs (hunk 1 above)
Comment (arein) [reply src/lib.rs]: How?... 🙏

File: src/stores/client.rs (hunk 3 above)
Comment (HarryET) [reply src/stores/client.rs:74]: Should work, but happy to merge with this and investigate later 😄

File: src/lib.rs (hunk 1 above)
Comment (HarryET) [reply src/lib.rs]: See `./providers/mod.rs` for examples using the `#[cfg()]` or here are some doc links: https://doc.rust-lang.org/rust-by-example/attribute/cfg.html, https://doc.rust-lang.org/reference/conditional-compilation.html

File: src/lib.rs [hunk 5]
```diff
@@ -27,6 +27,8 @@ use crate::{
     state::{Metrics, TenantStoreArc},
     stores::tenant::DefaultTenantStore,
 };
+
+#[cfg(not(any(debug_assertions, test)))]
```
Comment (HarryET) [inline src/lib.rs]: why?

File: src/lib.rs [hunk 6]
```diff
@@ -170,6 +172,7 @@ pub async fn bootstap(mut shutdown: broadcast::Receiver<()>, config: Config) ->
             },
         )
     } else {
+        #[cfg(not(any(debug_assertions, test)))]
```
Comment (HarryET) [inline src/lib.rs]: why?

**Review by HarryET — CHANGES_REQUESTED** [review]

File: src/lib.rs (hunk 5 above)
Comment (arein) [reply src/lib.rs]: We cannot have multiple instances of the server running with logging. It barfs about logging context if I try to do so.

Want me to add a comment?

File: src/lib.rs (hunk 6 above)
Comment (arein) [reply src/lib.rs]: Same as above

Comment (HarryET) [reply src/lib.rs]: Why did you not just use `else if cfg!() { /* ... */ }`

File: src/lib.rs (hunk 5 above)
Comment (HarryET) [reply src/lib.rs]: This disables them even just in debugging mode. Also see comment below

Comment (arein) [reply src/lib.rs]: I assumed this only disables when in `test`. Is there a way to do that?

File: src/lib.rs (hunk 6 above)
Comment (arein) [reply src/lib.rs]: I didn't know about that syntax.

File: src/error.rs [hunk 7]
```diff
@@ -88,6 +88,12 @@ pub enum Error {
 
     #[error("invalid tenant id: {0}")]
     InvalidTenantId(String),
+
+    #[error("server didn't boot within timeout")]
+    ServerBootTimeout(String),
+
+    #[error("server didn't shut down within timeout")]
+    ServerShutdownTimeout(String),
```
Comment (HarryET) [inline src/error.rs]: Thease should be removed. They aren't used.

File: src/lib.rs [hunk 8]
```diff
@@ -169,7 +171,8 @@ pub async fn bootstap(mut shutdown: broadcast::Receiver<()>, config: Config) ->
                 received_notifications: notification_counter,
             },
         )
-    } else {
+    } else if cfg!(not(test)) {
+        #[cfg(not(any(debug_assertions, test)))]
```
Comment (HarryET) [inline src/lib.rs]: This can be removed now that you have the `if cfg!()` syntax above

**Review by HarryET — CHANGES_REQUESTED** [review]

File: src/env.rs [hunk 9]
```diff
@@ -16,6 +16,8 @@ pub struct Config {
     pub tenant_database_url: Option<String>,
     #[serde(default = "default_tenant_id")]
     pub default_tenant_id: String,
+    #[serde(default = "default_is_test")]
+    pub is_test: bool,
```
Comment (HarryET) [inline src/env.rs:21]: Please use serde ignore so that people cannot send this as an env var and hide it at the bottom of the config struct

**Review by HarryET — CHANGES_REQUESTED** [review]

File: src/env.rs (hunk 9 above)
Comment (arein) [reply src/env.rs:21]: You mean `serde skip`? https://serde.rs/field-attrs.html#skip

Comment (HarryET) [reply src/env.rs:21]: yah

File: src/env.rs [hunk 10]
```diff
@@ -16,6 +16,8 @@ pub struct Config {
     pub tenant_database_url: Option<String>,
     #[serde(default = "default_tenant_id")]
     pub default_tenant_id: String,
+    #[serde(default = "default_is_test", skip)]
+    pub is_test: bool,
```
Comment (HarryET) [inline src/env.rs:21]:
```suggestion
    /// This is an internal flag to disable logging, cannot be defined by user
    pub is_test: bool,
```

**Review by HarryET — CHANGES_REQUESTED** [review]

**Pull request merged**